src/test/pybind/test_cephfs.py

```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+        cephfs.stat('dir1/dir3')
+
+        # cleanup
+        cephfs.chmod('/dir1/dir3', 0o755)
+        cephfs.rmtree('dir1', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir1')
+
+    def test_rmtree_no_perm_on_root(self, testdir):
+        '''
+        Test rmtree() exits when permission is not granted for the root of the file
+        hierarchy.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir1', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+
+        cephfs.chmod('/dir1', 0o000)
+        cephfs.rmtree('dir1', should_cancel)
+        # ensure /dir1 wasn't deleted
+        cephfs.stat('dir1')
+
+        # cleanup
+        cephfs.chmod('/dir1', 0o755)
+        cephfs.rmtree('dir1', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir1')
+
+    def test_rmtree_on_tree_with_snaps(self, testdir):
+        '''
+        Test that rmtree() successfully deletes the entire file hierarchy except
+        the branch where one of the directories contains one or many snapshots.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir1', 0o755)
+        cephfs.mkdir('dir1/dir2', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/dir2/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+        cephfs.mksnap('/dir1/dir2', 'snap1', 0o755)
+
+        cephfs.rmtree('dir1', should_cancel)
+        # ensure dir1 wasn't deleted
+        cephfs.stat('dir1')
+
+        # cleanup
+        cephfs.rmsnap('/dir1/dir2', 'snap1')
+        cephfs.rmtree('dir1', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir1')
+
+    def test_rmtree_on_tree_with_snaps_on_root(self, testdir):
+        '''
+        Test that rmtree() successfully deletes the entire file hierarchy except
+        the branch where one of the directories contains one or many snapshots.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir1', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+        cephfs.mksnap('/dir1', 'snap1', 0o755)
+
+        cephfs.rmtree('dir1', should_cancel)
+        # ensure dir1 wasn't deleted
+        cephfs.stat('dir1')
+
+        # cleanup
+        cephfs.rmsnap('/dir1', 'snap1')
+        cephfs.rmtree('dir1', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir1')
+
+    def test_rmtree_aborts_when_should_cancel_is_true(self, testdir):
+        '''
+        Test that rmtree() stops deleting the file hierarchy when the return
+        value of "should_cancel" becomes True.
+        '''
+        cancel_flag = False
+        should_cancel = lambda: cancel_flag
```

Comment:
To make this not racy, I suggest adding a sleep to this `should_cancel` procedure such that it would take more than 0.5 seconds to finish unlinking dir6.